### DISCUSSION TITLE
Energy thresholds in ausgab media definition

### DISCUSSION THREAD

**Discussion by Sasa-fras** [nrc-cnrc/EGSnrc#1113] [Questions] [answered]
I've already checked the section on this in the class library but I don't fully understand what it does. Do the upper and lower energy limits for the particles effect the creation of new particles from collisions? Or do they have some relation to the beam source? Or is it that if a particle from the beam source ever goes outside the thresholds it gets deleted and stops being tracked? I want to cutoff all electrons with energy less than 8MeV and have them stop being simulated and thus prevent them from impacting my energy deposition data. I was hoping this is the exact threshold I would use for that.

**Comment by ftessier** [comment, marked as answer]
In this case you want to set the `ECUT` (or `PCUT`) thresholds in the `transport parameters`.

---

In EGSnrc, there exist three sets of thresholds (or cutoffs): `AE, AP`; `UE, UP`; and `ECUT, PCUT`.

### `AE` and `AP`

`AE` represents the minimum electron energy in the simulation, while `AP` denotes the minimum photon energy. Note that these are _kinetic_ energies, with `AE` _including_ the electron rest mass (approximately 0.511 MeV).

**These thresholds determine particle production.** They establish the energy level below which secondary particles are _not explicitly created and tracked._ Instead, the energy lost to events that would produce particles below `AE` or `AP` contributes to the continuous energy loss of the incident particle, and is deposited locally. Interactions generating secondary particles _above_ these thresholds are modelled explicitly, with secondaries tracked in full detail.

During simulation initialization, EGSnrc pre-calculates data tables for each medium based on `AE` and `AP`. No data is generated below these energies, thus no particle can _exist_ in the simulation below `AE` and `AP`.

The minimum values in EGSnrc are 0.512 MeV for `AE` (1 keV kinetic) and 0.001 MeV for `AP`. Typical values are 0.521 and 0.010 (10 keV kinetic).

These parameters significantly impact simulation efficiency since lower thresholds result in the creation and tracking of numerous additional low-energy particles. A useful guideline is to begin with a kinetic energy of 10 keV and adjust accordingly for optimal simulation efficiency. If you can afford 1 keV, got for it!

A more refined rule of thumb is to set `AE` so that the range of electrons is about 1/3 of the smallest dimension in the geometry (so that at least a few electron interactions are likely to occur in the smallest region).

At any rate, it is highly recommended to test one representative simulation to a simulation with 1 keV kinetic threshold, to ensure that higher values of `AE` and `AP` are not affecting the final calculated value of interest.

### `UE` and `UP`

**These thresholds represent the maximum total energies** for which pre-calculated data tables are generated for each medium at the simulation's onset, for electrons and photons, respectively. Consequently, no particle can _exist_ in the simulation above these upper energies.

These thresholds have minimal impact on simulation outcome or efficiency. The only requirement is that they exceed the maximum energy of the source particles. It's advisable to set them not too much higher than the maximum source particle energy to ensure accurate interpolation of the generated data tables.

### `ECUT` and `PCUT`

**These thresholds serve as transport cutoff energies** for electrons and photons, respectively. When a particle's energy drops to this level as it loses energy during transport, it is immediately discarded, and the remaining energy is locally deposited. By default, the transport cutoffs are set to match `AE` and `AP`. However, increasing them, for instance, to halt electron or photon transport in specific cases, may be useful, as in the scenario you mention.

When `ECUT` exceeds the production thresholds `AE`, secondary electrons with energies between `AE` and `ECUT` are generated but then stopped immediately, with their energy deposited locally. This differs from merely setting `AE` to the higher value of `ECUT`, as it accurately models the straggling of the incident particle energy.

Note that if `ECUT` and `PCUT` are set below `AE` and `AP`, respectively, EGSnrc increases them to `AE` and `AP` upon initialization, as no particle can exist below `AE` and `AP` in the simulation.

### Summary

The foregoing explanation can be summarized visually as follows, in terms of the total energy of the secondary electron or photon:

![image](https://github.com/nrc-cnrc/EGSnrc/assets/1444241/25ae60ef-6d9e-4747-8b40-1c1095826a1e)

Here `csda` stands for _continuous slowing down approximation_: in that range secondaries are not created at all.

**Reply by Sasa-fras** [reply]
Thanks, that helps a lot and was really informative and exceptionally well organized for a reply. **_I think there would still be an issue for me_** though. In my case I am only interested in energy deposited from collisions resultant from an incoming electron with at least 8MeV. I would want a cut off at 8MeV so that energy deposited from collisions with **_electrons less than 8MeV don't contribute to the data_**. It sounds like _**if I set a cut off energy then it would deposit all the remaining energy locally which would be even worse**_ then letting the electrons continue to collide with stuff. **_Is there a way I could use ECUT but have the energy disappear instead of being deposited locally?_** Thanks again for being so helpful, I wish it was easier to find this information so I wouldn't be bothering you.

**Reply by mchamberland** [reply]
@Sasa-fras I think if you were to set the particle weight to zero, it would be discarded without depositing energy. Is that right, @ftessier?

**Reply by ftessier** [reply]
I see. There may be a way to do this with the thresholds, but I cannot think of one! The more general answer, and more flexible option, is to implement that in the code itself. As @mchamberland said, this can be achieved by setting the particle weight to 0 when its energy drops to 8 MeV.

If you don't need any advanced variance reduction techniques, you can do this in the ausgab function in the small demo application egs_app. You can also do this in any application, but in some cases the ausgab function becomes quite involved. You will likely want to insert the code at the beginning of ausgab, at any rate.

I will post the corresponding c++ code snippet here when I have a(nother) minute!

**Reply by Sasa-fras** [reply]
Couldn't there be problems/repercussions from setting the weight to zero? I don't know what that property is supposed to do in the code, but surely if it was properly simulated as if its mass was zero then that would screw a lot of stuff up. Or is weight more of a scaling modifier type of weight rather than mass. Since after all, even if it had a mass of zero it would still have kinetic energy to deposit if i cut it off at 8MeV.

**Reply by ftessier** [reply]
Sorry for the confusion, we are not talking about mass, but "statistical weight" here. Setting the statistical weight to zero discards the particle. Which EGSnrc application are you using for your project?

**Reply by Sasa-fras** [reply]
The default one, its called egs_app.

**Reply by ftessier** [reply]
Ah ok, perfect, I can post the code lines shortly. Do you know c++ a little? You don't have to, but I want to know what I shall explain... 😄

**Reply by Sasa-fras** [reply]
I only know basic syntax, took an intro course back in high school. I mostly work with python.

**Reply by ftessier** [reply]
All right, so most of what you need is **already there** in the `egs_app.cpp` source file:

```c++
// ausgab
int myapp::ausgab(int iarg) {

    // stack index variables
    int np  = the_stack->np - 1;
    int npold  = the_stack->npold - 1;
    int region = the_stack->ir[np]-2;

    // stack phase space variables
    double E = the_stack->E[np];
    double x = the_stack->x[np];
    double y = the_stack->y[np];
    double z = the_stack->z[np];
    double u = the_stack->u[np];
    double v = the_stack->v[np];
    double w = the_stack->w[np];
    double weight = the_stack->wt[np];
    int iq = the_stack->iq[np];

    // latch stack variable
    int latch  = the_stack->latch[np];

    // handle the ausgab call

    return 0;
}
```

The `ausgab` callback function is called whenever anything happens to particles in the underlying EGSnrc engine. You simply need to add the following code under `// handle the ausgab call`:

```c++

    // handle the ausgab call
    if (E < 8.0) {
        the_stack->wt[np] = 0.0;
    }
```

(or 8.511 if you meant 8 MeV _kinetic_ energy for electrons). You can add `&& iq == -1` in the `if (...)` conditional to discard only electrons (or `iq == 0` for photons, `iq == 1` for positrons). You can consider adding more code, for example, to track the total energy that _disappears_ this way! Let me know if you need further help.

Once you have saved the file, issue the command `make` in the `egs_app/` directory, which will recompile the application. Then, run `egs_app` as usual, and let me know how it pans out!

**Reply by Sasa-fras** [reply]
I added it and confirmed a drop in energy deposition, thanks for the overwhelming support.

**Reply by ftessier** [reply]
Another option, that would leave the physics unaffected, is to _score_ the energy of the electron in
`ausgab()` when it is larger than 8 MeV, but otherwise let it continue its journey. That involves creating a new variable and a corresponding output statement in `outputResults()`. Or else you can score the electron energy spectrum, and post-process the data to select and integrate the energy range of interest.
